srtcore/core.cpp

```diff
@@ -10552,6 +10620,42 @@ int srt::CUDT::processData(CUnit* in_unit)
         return -1;
     }
 
+    // 1. This is set to true in case when TSBPD during the last check
+    // has seen no packet candidate to ever deliver, hence it needs
+    // an update on that. Note that this is also false if TSBPD thread
+    // isn't running.
+    // 2. If next_tsbpd_avail is set, it means that in the buffer there is
+    // a new packet that precedes the previously earliest available packet.
+    // This means that if TSBPD was sleeping up to the time of this earliest
+    // delivery (after drop), this time we have received a packet to be delivered
+    // earlier than that, so we need to notify TSBPD immediately so that it
+    // updates this itself, not sleep until the previously set time.
+
+    // The meaning of m_bWakeOnRecv:
+    // - m_bWakeOnRecv is set by TSBPD thread and means that it wishes to be woken up
+    //   on every received packet. Hence we signal always if a new packet was inserted.
+    // - even if TSBPD doesn't wish to be woken up on every reception (because it sleeps
+    //   until the play time of the next deliverable packet), it will be woken up when
+    //   next_tsbpd_avail is set because it means this time is earlier than the time until
+    //   which TSBPD sleeps, so it must be woken up prematurely. It might be more performant
+    //   to simply update the sleeping end time of TSBPD, but there's no way to do it, so
+    //   we simply wake TSBPD up and count on that it will update its sleeping settings.
+
+    //   XXX Consider: as CUniqueSync locks m_RecvLock, it means that the next instruction
+    //   gets run only when TSBPD falls asleep again. Might be a good idea to record the
+    //   TSBPD end sleeping time - as an alternative to m_bWakeOnRecv - and after locking
+    //   a mutex check this time again and compare it against next_tsbpd_avail; might be
+    //   that if this difference is smaller than "dirac" (could be hard to reliably compare
+    //   this time, unless it's set from this very value), there's no need to wake the TSBPD
+    //   thread because it will wake up on time requirement at the right time anyway.
+    if (m_bTsbPd && ((m_bWakeOnRecv && new_inserted) || next_tsbpd_avail != time_point()))
+    {
+        HLOGC(qrlog.Debug, log << "processData: will SIGNAL TSBPD for socket. WakeOnRecv=" << m_bWakeOnRecv
+                << " new_inserted=" << new_inserted << " next_tsbpd_avail=" << FormatTime(next_tsbpd_avail));
+        CUniqueSync tsbpd_cc(m_RecvLock, m_RcvTsbPdCond);
+        tsbpd_cc.notify_all();
+    }
+
```

Comment:
Might be that `m_bWakeOnRecv` still has to be protected by `m_RecvLock` to ensure the TSBPD thread has set it appropriately before it is checked here. Making it atomic does not resolve this transition state.

Reply:
How about:

```suggestion
    if (m_bTsbPd)
    {
        CUniqueSync tsbpd_cc(m_RecvLock, m_RcvTsbPdCond);
        if ((m_bWakeOnRecv && new_inserted) || next_tsbpd_avail != time_point())
        {
            HLOGC(qrlog.Debug, log << "processData: will SIGNAL TSBPD for socket. WakeOnRecv=" << m_bWakeOnRecv
                    << " new_inserted=" << new_inserted << " next_tsbpd_avail=" << FormatTime(next_tsbpd_avail));
            tsbpd_cc.notify_all();
        }
    }
```